File: OnBoard/Components/PduR/Implementation/PduR.c
#include "PduR.h"
#include "PduR_CanTp.h"
#include "PduR_Com.h"
#include "PduR_Cfg.h"

#include "Com.h"
#include "CanTp.h"
#include "Dio.h"

extern const PduR_ConfigType PduR_PCConfig;

static PduR_StateType PduR_State = PDUR_UNINIT;

static const PduR_ConfigType *PduR_ConfigPtr = (const PduR_ConfigType *)0;

void PduR_Init(const PduR_ConfigType *ConfigPtr)
{
#if (STD_ON == PDUR_PRECOMPILE_SUPPORT)
    PduR_ConfigPtr = &PduR_PCConfig;
    (void)ConfigPtr;
#else
    PduR_ConfigPtr = ConfigPtr;
#endif

    if ((PduR_ConfigPtr->Routes == (const PduR_RouteType *)0) ||
        (PduR_ConfigPtr->RouteCount == 0u))
    {
        PduR_State = PDUR_UNINIT;
        return;
    }

    PduR_State = PDUR_ONLINE;
}

Std_ReturnType PduR_ComTransmit(PduIdType TxPduId,
                                const PduInfoType *PduInfoPtr)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR) ||
        (PduInfoPtr == NULL_PTR))
    {
        return E_NOT_OK;
    }

    if ((PduInfoPtr->SduLength > 0U) &&
        (PduInfoPtr->SduDataPtr == NULL_PTR))
    {
        return E_NOT_OK;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].SrcPduId == TxPduId)
        {
            return CanTp_Transmit(
                PduR_ConfigPtr->Routes[i].DstPduId,
                PduInfoPtr);
        }
    }

    return E_NOT_OK;
}

void PduR_CanTpRxIndication(PduIdType RxPduId,
                            Std_ReturnType result)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR))
    {
        return;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].DstPduId == RxPduId)
        {
            // Dio_WriteChannel(DioConf_DioChannel_DioChannel_LedBlue, STD_HIGH);
            /* Forward lên COM bằng SrcPduId */
            Com_TpRxIndication(PduR_ConfigPtr->Routes[i].SrcPduId,
                               result);

            break;
        }
    }
}

void PduR_CanTpTxConfirmation(PduIdType TxPduId,
                              Std_ReturnType result)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR))
    {
        return;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].DstPduId == TxPduId)
        {
            /* Forward lên COM bằng SrcPduId */
            Com_TpTxConfirmation(PduR_ConfigPtr->Routes[i].SrcPduId,
                                 result);

            break;
        }
    }
}

BufReq_ReturnType PduR_CanTpStartOfReception(
    PduIdType id,
    const PduInfoType *info,
    PduLengthType TpSduLength,
    PduLengthType *bufferSizePtr)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR) ||
        (bufferSizePtr == NULL_PTR))
    {
        return BUFREQ_E_NOT_OK;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].SrcPduId == id)
        {
            /* Forward lên COM bằng SrcPduId */
            return Com_StartOfReception(
                PduR_ConfigPtr->Routes[i].DstPduId,
                info,
                TpSduLength,
                bufferSizePtr);
        }
    }

    return BUFREQ_E_NOT_OK;
}

BufReq_ReturnType PduR_CanTpCopyRxData(PduIdType id, const PduInfoType *info, PduLengthType *bufferSizePtr)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR) ||
        (bufferSizePtr == NULL_PTR))
    {
        return BUFREQ_E_NOT_OK;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].DstPduId == id)
        {
            /* Forward lên COM bằng SrcPduId */
            return Com_CopyRxData(
                PduR_ConfigPtr->Routes[i].SrcPduId,
                info,
                bufferSizePtr);
        }
    }

    return BUFREQ_E_NOT_OK;
}

BufReq_ReturnType PduR_CanTpCopyTxData(PduIdType id, const PduInfoType *info, const RetryInfoType *retry, PduLengthType *availableDataPtr)
{
    if ((PduR_State != PDUR_ONLINE) ||
        (PduR_ConfigPtr == NULL_PTR) ||
        (availableDataPtr == NULL_PTR))
    {
        return BUFREQ_E_NOT_OK;
    }

    for (uint16 i = 0U; i < PduR_ConfigPtr->RouteCount; i++)
    {
        if (PduR_ConfigPtr->Routes[i].DstPduId == id)
        {
            /* Forward lên COM bằng SrcPduId */
            return Com_CopyTxData(
                PduR_ConfigPtr->Routes[i].SrcPduId,
                info,
                retry,
                availableDataPtr);
        }
    }

    return BUFREQ_E_NOT_OK;
}

File: OnHost/bus_monitor/Components/PduR/PduR.py
from __future__ import annotations

from typing import Optional

from Components.PduR.PduR_Types import (
    PduR_ConfigType,
    PduR_StateType,
    PduR_ModuleType,
)
from Components.PduR.PduR_Cfg import PduR_PCConfig


class PduR:
    def __init__(self, com=None, cantp=None) -> None:
        self._com = com
        self._cantp = cantp

        self._PduR_State = PduR_StateType.PDUR_UNINIT
        self._PduR_ConfigPtr: Optional[PduR_ConfigType] = None

    def set_com(self, com) -> None:
        self._com = com

    def set_cantp(self, cantp) -> None:
        self._cantp = cantp

    def PduR_Init(self, ConfigPtr: Optional[PduR_ConfigType] = None) -> None:
        self._PduR_ConfigPtr = PduR_PCConfig if ConfigPtr is None else ConfigPtr

        if self._PduR_ConfigPtr.Routes is None or self._PduR_ConfigPtr.RouteCount == 0:
            self._PduR_State = PduR_StateType.PDUR_UNINIT
            return

        self._PduR_State = PduR_StateType.PDUR_ONLINE

    def PduR_ComTransmit(self, TxPduId: int, payload: bytes) -> bool:
        if (
            self._PduR_State != PduR_StateType.PDUR_ONLINE
            or self._PduR_ConfigPtr is None
            or payload is None
        ):
            return False

        if len(payload) > 0 and payload is None:
            return False

        for route in self._PduR_ConfigPtr.Routes:
            if route.SrcPduId != TxPduId:
                continue

            if route.SrcModule != PduR_ModuleType.PDUR_COM:
                continue

            if route.DstModule != PduR_ModuleType.PDUR_CANTP:
                continue

            if self._cantp is None:
                return False

            return self._cantp.CanTp_Transmit(route.DstPduId, payload)

        return False

    def PduR_CanTpRxIndication(
        self,
        RxPduId: int,
        payload: bytes,
        result: bool,
    ) -> None:
        if (
            self._PduR_State != PduR_StateType.PDUR_ONLINE
            or self._PduR_ConfigPtr is None
        ):
            return

        for route in self._PduR_ConfigPtr.Routes:
            if route.SrcModule != PduR_ModuleType.PDUR_CANTP:
                continue

            if route.DstModule != PduR_ModuleType.PDUR_COM:
                continue

            if route.SrcPduId == RxPduId:
                if self._com is not None:
                    self._com.Com_RxIndication(route.DstPduId, payload, result)
                return

    def PduR_CanTpTxConfirmation(self, TxPduId: int, result: bool) -> None:
        if (
            self._PduR_State != PduR_StateType.PDUR_ONLINE
            or self._PduR_ConfigPtr is None
        ):
            return

        for route in self._PduR_ConfigPtr.Routes:
            if route.SrcModule != PduR_ModuleType.PDUR_COM:
                continue

            if route.DstModule != PduR_ModuleType.PDUR_CANTP:
                continue

            if route.DstPduId == TxPduId:
                if self._com is not None:
                    self._com.Com_TxConfirmation(route.SrcPduId, result)
                return
